Parallel blocks (#31)
CQCL/phir: spec.md

- `"format"`: Signifies the utilization of the PHIR/JSON format.
- `"version"`: Represents the semantic version number adhering to the PHIR spec.
- `"metadata"`: An optional segment where users can incorporate additional details. This segment holds potential for
future expansion, possibly to guide compilation processes and error modeling.
- `"ops": [{...}, ...]`: A linear sequence denoting the operations and blocks that constitute the program.

## Metadata

> The metadata field takes several parameters that can be used to specify information relevant to the entire program

Reviewer: This description feels somewhat inconsistent with the description above. We should mention that the following table shows possible metadata and should include other fields used in the examples in this spec.

### Metadata Options

| parameter               | options                 | description                                                     |
| ----------------------- | ----------------------- | --------------------------------------------------------------- |
| `"default_time_unit"`   | `"s", "ms", "us", "ns"` | The default time unit in the "duration" section of qop metadata |

> | `"default_angle_unit"`  | `"rad", "pi"`           | The default angle unit in the "angles" section of a qop         |

Reviewer: "default_" is not needed as a prefix. The second could be:
```suggestion
| `"angular_unit"`  | `"rad", "pi"`           | The default angle unit in the "angles" section of a qop         |
```

But also if we are adding these two, we will need to revise the rest of the spec. I suggest not introducing these two units in this PR to keep the focus of this PR toward parallelism only.

> | `"strict_parallelism"`  | `"true", "false"`       | If true, interperet qops with multiple qubits outside a parallel block as the same gate being executed in parallel on all qubits. If false, interperet qops with multiple qubits outside a parallel block as the same gate being executed serially on each qubit in turn. | {{< mdl-disable "<!-- markdownlint-disable MD013 -->" >}}

Reviewer: ```suggestion
| `"strict_parallelism"`  | `"true", "false"`       | If `"true"`, interpret `qop`s with multiple arguments as parallel application of the `qop` to those arguments. If `"false"` (the default), interpret qops with multiple qubits outside a parallel block as the same gate being executed serially on each qubit in turn. | {{< mdl-disable "<!-- markdownlint-disable MD013 -->" >}}
```

Didn't we say that `false` implies that the simulator gets to choose? It doesn't mean strictly serial, right?

}
```

### Parallel quantum block

A grouping of quantum operations to be performed in parallel

```json5
{
  "block": "parallel",
  "ops": [{...}, ...],
  "metadata": {...}  // Optional
}

> ```

Reviewer: We should add a note here that only quantum operations are allowed in the `"ops"` field here.

I'd also like to see whether @qciaran thinks the block name should be called "qparallel" or some such name to be clear that this is only relevant for quantum operations.

Reviewer: @Asa-Kosto-QTM unsure if this was missed. We restrict the ops to be `QOp` in the model, so the spec should clearly say that. General `"ops"` allow blocks and all kinds of ops within them.

}
```

### Parallel quantum block

A grouping of quantum operations to be performed in parallel

```json5
{
  "block": "parallel",
  "ops": [{...}, ...],
  "metadata": {...}  // Optional
}
```

> #### Parallel quantum block example

Reviewer: ```suggestion
For example:
```

```json5
{
  "block": "parallel",
  "ops": [{"qop": "RZ", "angles": [[1.5], "pi"], "args": [["q", 0], ["q", 1]]},
          {"qop": "RZ", "angles": [[1.0], "pi"], "args": [["q", 2], ["q", 3]]},
          {"qop": "RZ", "angles": [[0.5], "pi"], "args": [["q", 4], ["q", 5]]}
          ]
}
```

> 

Reviewer: Also, add a note either here or while introducing the example before the code block such as: "the six gate applications here are interpreted to be applied in parallel".

}
```

### Parallel block

> A grouping of to be performed in parallel. Currently only quantum operations are supported.

Reviewer 2: Assuming we are doing "qparallel" instead. Maybe instead just say: "A grouping of quantum operations to be performed in parallel."

```json5
{

>   "block": "parallel",

Reviewer 2: ```suggestion
  "block": "qparallel",
```

  "ops": [{...}, ...],
  "metadata": {...}  // Optional
}
```

This example contains 6 RZ gate applications. There is 1 `"qop"` per unique gate angle, each with 2 qubit arguments.
All gates within the block will be applied in parallel.

```json5
{

>   "block": "parallel",

Reviewer 2: ```suggestion
  "block": "qparallel",
```

}
```

### Parallel quantum block

A grouping of quantum operations to be performed in parallel

```json5
{
  "block": "parallel",
  "ops": [{...}, ...],
  "metadata": {...}  // Optional
}
```

> Example:

Reviewer: ```suggestion
```

> This example contains 6 RZ gates. There is 1 "qop" per unique gate angle, and 2 arguments.

Reviewer: ```suggestion
The following example contains 6 RZ gate applications. There is 1 `"qop"` per unique gate angle each with 2 qubit arguments.
```

}
```

### Parallel block

> A grouping of quantum operations to be performed in parallel

Reviewer: See comments in the `model.py`. If this block is for quantum ops only, the name should reflect that.
